File: hw4/loans.py
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

def tidy_loans(df):
    non_status_fields = ['id', 'account_id', 'date', 'amount', 'payments']
    status_fields = [e for e in df.columns.tolist() if e not in non_status_fields]
    binary_status_fields = df[status_fields].replace('X', 1).replace('-', 0)
    loans2 = pd.concat([df[non_status_fields], binary_status_fields],axis=1)
    melted = loans2.melt(id_vars=non_status_fields)
    loans3 = melted[melted['value'] == 1]
    loans3['term'] = loans3['variable'].apply(lambda x: int(x.split("_")[0]))
    loans3['status'] = loans3['variable'].apply(lambda x: x.split("_")[1])
    loans3['active_status'] = loans3['status'].apply(lambda x: 'expired' if x in ['A','B'] else 'current')
    loans3['payment'] = loans3['status'].apply(lambda x: 'full' if x == 'A' else ('in_progress' if x == 'C' else 'default'))
    loans4 = loans3.drop(['variable', 'value','status'], axis=1)
    loans4.reset_index(drop=True, inplace=True)
    loans4.to_csv('loans_py.csv')
    return loans4

if __name__ == '__main__':
    loans = pd.read_csv('data/loans.csv')
    tidy_loans(loans)

File: hw4/loans.R
library(magrittr) 
library(dplyr) 
library(tidyverse)
library(reshape)
options(warn=-1)

tidy_loans <- function(loans){
  non_status_fields = c('id', 'account_id', 'date', 'amount', 'payments')
  status_fields = setdiff(colnames(loans), non_status_fields)
  binary_status_fields = loans %>% select(status_fields)
  new_binary <- binary_status_fields %>% 
    mutate_all(funs(str_replace(., "X", "1"))) %>%
    mutate_all(funs(str_replace(., "-", "0")))
  loans2 <- cbind(loans %>% select(non_status_fields), as.data.frame(sapply(new_binary, as.numeric)) )
  loans3 <- melt(loans2, id = non_status_fields) %>% filter(value == 1)
  loans4 <- loans3 %>%
    separate(variable, c("term", "status"), "_") %>%
    separate(term, c("todrop", "term"), "X") 
  loans4$active_status <- ifelse(loans4$status %in% c("A", "B") , "expired", "current")
  loans4$payment <- ifelse(loans4$status %in% c("B", "D"), "default", ifelse(loans4$status == "A", "full", "in_progress"))
  loans4 %>% select(c(non_status_fields, "term", "active_status", "payment"))
  write.csv(loans4, "loans_r.csv")
  return(loans4)
}

main <- function(){
  loans = read.csv('./data/loans.csv')
  tidy_loans(loans)
}

if (!interactive()) {
  main()
}